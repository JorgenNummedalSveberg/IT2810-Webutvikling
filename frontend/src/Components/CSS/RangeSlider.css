.RangeSlider{
    width: 100%;
    background-color: rgb(250, 250, 250);
    border-radius: 8px;
    padding: 10px;
    margin-bottom: 20px;
    -webkit-box-shadow: 0px 4px 4px 0px rgba(0,0,0,0.25); 
    box-shadow: 0px 4px 4px 0px rgba(0,0,0,0.25);
}
.RangeSlider h2{
    font-size: 1.5em;
    width: 100%;
}
