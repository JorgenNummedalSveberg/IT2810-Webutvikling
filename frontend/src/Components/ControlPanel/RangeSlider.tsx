import React, {useRef} from 'react';
import {Paper, Slider} from '@material-ui/core';
import {useDispatch} from "react-redux";
import {setPage, setScore, setYears} from '../../actions';
import {makeStyles} from '@material-ui/styles';
import ImdbIcon from "../Shared/ImdbIcon";


export default function RangeSlider(props: { score: number[], type: string, refresh: () => void }) {

    // Value som setter verdien på slidern
    const [value, setValue] = React.useState<number[]>([props.score[0], props.score[1]]);
    const [range] = React.useState<number[]>(props.type === "year" ? [1900, 2020] : [0, 10]);


    // Nødvendig for redux
    const dispatch = useDispatch();

    // Tom timeout ref som defineres først;
    let timeoutRef = useRef(setTimeout(() => {
    }, 0));

    // Når input endres tømmer vi den aktive timeouten og starter på nytt. Når der har gått 300ms, oppdater score range
    function handleChange(e: any, data: number | number[]) {
        clearTimeout(timeoutRef.current);
        setValue(data as number[]);
        timeoutRef.current = setTimeout(() => {
            dispatch(setPage(0));
            if (props.type === 'score') {
                dispatch(setScore(data as number[]));
            } else {
                dispatch(setYears(data as number[]));
            }
            props.refresh();
        }, 300);
    }

    const classes = makeStyles({
        slider: {
            backgroundColor: '#E98074',
            padding: '20px',
            borderRadius: '10px'
        },
        sliderLabel: {
            display: 'flex',
            justifyContent: 'space-between'
        },
        imdbLogo: {
            marginTop: '20px'
        }
    })
    return (
        <Paper className={classes().slider}>
            <Slider
                value={value}
                onChange={handleChange}
                valueLabelDisplay="auto"
                step={props.type === 'year' ? 5 : 0.5}
                marks
                min={range[0]}
                max={range[1]}
            />
            <div className={classes().sliderLabel}>
                {props.type === 'year' ?
                    <h3>{value[0]}</h3> :
                    <div className={classes().imdbLogo}><ImdbIcon rating={value[0]} height={25}/></div>}
                <h2>{props.type === 'year' ? "Year" : "Rating"}</h2>
                {props.type === 'year' ?
                    <h3>{value[1]}</h3> :
                    <div className={classes().imdbLogo}><ImdbIcon rating={value[1]} height={25}/></div>}
            </div>
        </Paper>
    )
}