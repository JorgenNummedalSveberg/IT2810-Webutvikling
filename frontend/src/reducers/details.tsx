// Reducer for lagring av filmen som skal vises i popup
import {Movie} from "../types/Movie";

const movieReducer = (details = {show: false, movie: {}}, action: { type: string; payload: {show: boolean, movie: Movie }; }) => {
    switch (action.type) {
        case 'setMovie':
            return action.payload;
        default:
            return details;
    }
}
export default movieReducer;