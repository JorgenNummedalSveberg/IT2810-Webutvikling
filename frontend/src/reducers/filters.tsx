// Samlet reducer for alle filterene
const filterReducer = (filters = {desc:true, genre: "", search: "", sort: "", score: [0,10], year: [1900,2020]}, action: { type: string; payload: string | boolean | number[]; }) => {
    let result = filters;
    switch (action.type) {
        // Endrer retning vi sorterer i
        case 'setDesc':
            result.desc = action.payload as boolean;
            break;
        // Endrer tittel søkekriterie
        case 'setSearch':
            result.search = action.payload as string;
            break;
        // Endrer sjanger søkekriterie
        case 'setGenre':
            result.genre = action.payload as string;
            break;
        // Endrer hva vi sorterer etter
        case 'setSort':
            result.sort = action.payload as string;
            break;
        // Endrer rating søkekriterie
        case 'setScore':
            result.score = action.payload as number[];
            break;
        // Endrer årstall søkekriterie
        case 'setYears':
            result.year = action.payload as number[];
            break;
    }
    return result;
}
export default filterReducer;