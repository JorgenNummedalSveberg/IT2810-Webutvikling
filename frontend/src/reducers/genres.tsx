// Reducer for Ã¥ endre sjangerlisten, setter en generisk liste som default
const genresReducer = (genres = [
    'Select genre...',
    'Biography', 'Drama',
    'History', 'Crime',
    'Fantasy', 'Mystery',
    'Comedy', 'Romance',
    'Family', 'Sci-Fi',
    'Adventure', 'Action',
    'Thriller', 'War',
    'Horror', 'Western',
    'Animation', 'Musical',
    'Sport', 'Music'
], action: { type: any; payload: any; }) => {
    switch (action.type) {
        case 'setGenres':
            return genres.concat(action.payload).sort();
        default:
            return genres;
    }
}
export default genresReducer;