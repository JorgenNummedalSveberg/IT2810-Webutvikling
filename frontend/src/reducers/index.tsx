import moviesReducer from "./movies";
import genresReducer from "./genres";
import {combineReducers} from "redux";
import filterReducer from "./filters";


// Lager en samlet reducer slik at alt kan lagres i en enkelt store
const reducers = combineReducers({
    movies: moviesReducer,
    genres: genresReducer,
    filter: filterReducer
})

export default reducers;