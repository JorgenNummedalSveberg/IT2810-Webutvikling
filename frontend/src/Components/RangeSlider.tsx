import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Slider from '@material-ui/core/Slider';
import './CSS/RangeSlider.css';
import {state} from "../types/state";
import {useDispatch, useSelector} from "react-redux";
import { setScore } from '../actions';

const useStyles = makeStyles({
  root: {
    width: 200,
  },    
});

function valuetext(value: number) {
  return `${value}°C`;
}

export default function RangeSlider() {
    
    // Nødvendig for redux
    const dispatch = useDispatch();

    // Henter inn sjangre fra redux state
    const minScore = useSelector((state: state) => state.filter.minScore);

    const [value, setValue] = React.useState<number[]>([minScore[0], minScore[1]]);



    const handleChange = (event: any, newValue: number | number[]) => {
        setValue(newValue as number[]);
        dispatch(setScore(newValue as number[]));
        //Kan legge til at det automatisk skal byttes til rating sortering
        //dispatch(setSort("Rating"));
    };

  return (
    <div className={"RangeSlider"}>
        <h2>Minimum score</h2>
        <Slider className="Slider"
            value={value}
            onChange={handleChange}
            valueLabelDisplay="auto"
            aria-labelledby="range-slider"
            getAriaValueText={valuetext}
        />
    </div>
  );
}