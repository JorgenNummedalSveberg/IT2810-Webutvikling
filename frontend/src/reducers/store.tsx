import {createStore} from "redux";
import reducers from "../reducers";

// Store for redux, @ts-ignore delen er for Ã¥ kunne se redux i redux webtools
export const store = createStore(
    reducers,
    // @ts-ignore
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)