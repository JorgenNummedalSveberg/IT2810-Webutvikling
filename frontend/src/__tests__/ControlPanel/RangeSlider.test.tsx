import React from 'react';
import {render} from '@testing-library/react';
import RangeSlider from "../../Components/ControlPanel/RangeSlider";
import {Provider} from "react-redux";
import {store} from "../../reducers/store";

describe("Test to see if the sort button renders with our mock function, and passes props correctly", () => {
    test("Rendering of button", () => {
        const {container} = render(<Provider store={store}><RangeSlider score={[10, 100]} type={"year"}/></Provider>);
        expect(container).toMatchSnapshot();
    })
})