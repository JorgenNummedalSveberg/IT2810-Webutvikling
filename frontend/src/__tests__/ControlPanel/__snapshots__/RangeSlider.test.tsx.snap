// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to see if the sort button renders with our mock function, and passes props correctly Rendering of button 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-slider-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <span
      class="MuiSlider-root MuiSlider-colorPrimary makeStyles-sliderRange-4"
    >
      <span
        class="MuiSlider-rail"
      />
      <span
        class="MuiSlider-track"
        style="left: 0%; width: 0%;"
      />
      <input
        type="hidden"
        value="1900,1900"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="0"
        style="left: 0%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="1"
        style="left: 4.166666666666667%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="2"
        style="left: 8.333333333333334%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="3"
        style="left: 12.5%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="4"
        style="left: 16.666666666666668%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="5"
        style="left: 20.833333333333332%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="6"
        style="left: 25%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="7"
        style="left: 29.166666666666668%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="8"
        style="left: 33.333333333333336%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="9"
        style="left: 37.5%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="10"
        style="left: 41.666666666666664%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="11"
        style="left: 45.833333333333336%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="12"
        style="left: 50%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="13"
        style="left: 54.166666666666664%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="14"
        style="left: 58.333333333333336%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="15"
        style="left: 62.5%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="16"
        style="left: 66.66666666666667%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="17"
        style="left: 70.83333333333333%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="18"
        style="left: 75%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="19"
        style="left: 79.16666666666667%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="20"
        style="left: 83.33333333333333%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="21"
        style="left: 87.5%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="22"
        style="left: 91.66666666666667%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="23"
        style="left: 95.83333333333333%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="24"
        style="left: 100%;"
      />
      <span
        aria-orientation="horizontal"
        aria-valuemax="2020"
        aria-valuemin="1900"
        aria-valuenow="1900"
        class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-5"
        data-index="0"
        role="slider"
        style="left: 0%;"
        tabindex="0"
      >
        <span
          class="PrivateValueLabel-offset-7 MuiSlider-valueLabel"
        >
          <span
            class="PrivateValueLabel-circle-8"
          >
            <span
              class="PrivateValueLabel-label-9"
            >
              1900
            </span>
          </span>
        </span>
      </span>
      <span
        aria-orientation="horizontal"
        aria-valuemax="2020"
        aria-valuemin="1900"
        aria-valuenow="1900"
        class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-5"
        data-index="1"
        role="slider"
        style="left: 0%;"
        tabindex="0"
      >
        <span
          class="PrivateValueLabel-offset-7 MuiSlider-valueLabel"
        >
          <span
            class="PrivateValueLabel-circle-8"
          >
            <span
              class="PrivateValueLabel-label-9"
            >
              1900
            </span>
          </span>
        </span>
      </span>
    </span>
    <div
      class="makeStyles-sliderLabel-2"
    >
      <h3>
        10
      </h3>
      <h2>
        Year
      </h2>
      <h3>
        100
      </h3>
    </div>
  </div>
</div>
`;
